### ISSUE TITLE
Update Android targetSdkVersion to 28, by 2019-11 deadline

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3563]
We should update our `targetSdkVersion` to 28, aka Android 9.0 Pie. It's currently 26, since #3075.

Starting 2019-11-01, Google Play will require this for any new updates to our app.

Upstream doc here:
https://developer.android.com/about/versions/pie/android-9.0-migration
and some useful background here:
https://medium.com/androiddevelopers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd

The important steps are
1. Read about the potentially-breaking changes, and identify those that might affect us.
2. Make a WIP change to `targetSdkVersion`.
3. Test the WIP change thoroughly, especially in the areas highlighted in step 1. Fix things as needed, and repeat.

Fortunately this is a less overwhelming set of changes than we had to deal with in #3075 -- because we aren't catching up on 4 years of platform changes at once, as we were then.

More details to fill in later. Filing now in order to have an issue to link to.

**Comment by gnprice** [comment]
This can also be seen as a followup task from the RN v0.59 upgrade #3399 -- the upstream RN template app made this change in that interval.

**Comment by gnprice** [comment]
I was just going through related docs to close out #3075 (and disaggregate the followup tasks that remain from it), so here are some more links.

A short outline of important changes in 9.0 / 28:
https://developer.android.com/distribute/best-practices/develop/target-sdk#prepie
That includes changes that already apply to us on 9.0, in addition to changes that are opt-in with `targetSdkVersion`.

A more detailed list, and focused exclusively on changes relevant to this issue (i.e., changes that will come into effect when we update `targetSdkVersion`):
https://developer.android.com/about/versions/pie/android-9.0-changes-28

A quick skim of that list finds:
* Likely requires action: `isCleartextTrafficPermitted` -- see also 1f55c035a
* Might require action: WebView changes re MIME types and re document scrolling
* Several others I'd have to investigate more to be confident they don't affect us.

Of the changes that already apply to us on Android 9 devices:
* [App Standby Buckets](https://developer.android.com/about/versions/pie/power#buckets) can affect ability to receive FCM messages (and hence show notifications.)
  * Confirms the importance of not overusing `priority: high` in FCM! Fortunately we've always gotten that right; see #3185.

**Comment by gnprice** [comment]
Rereading [that list](https://developer.android.com/about/versions/pie/android-9.0-changes-28):

* (Again the `isCleartextTrafficPermitted` change.)

* > Might require action: WebView changes re MIME types

  This applies to `file:` URLs. Currently our `baseUrl` is from `auth.realm`, so it's not on `file:`, and this doesn't affect us. Soon we'll make it a `file:` URL, with #3636 . I think this just means we need to make sure it has a `.html` extension when we do -- which that PR already does.

* > Might require action: WebView changes [...] re document scrolling

  Described in a couple of paragraphs [here](https://developer.android.com/about/versions/pie/android-9.0-changes-28#scrolling-element). This calls for further investigation; I know we use the properties this discusses, so we'll want to check our use against what this says is the new semantics.

* "Prior to Android 9, notifications from suspended apps were canceled. Beginning with Android 9, notifications from suspended apps are hidden until the app is resumed."

  I'm not sure what a "suspended app" is -- best I can find on the web for searches like [android suspend app] is about an app getting suspended *from the Play Store*, which doesn't sound like it'd be what this means. Rather, this sounds like something between being in the background and outright stopped?

  Probably OK to ignore this one.

* "The [ConnectivityManager.NetworkCallback](https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback.html) class now sends information about VPNs to apps. This change makes it much easier for apps to listen for connectivity events without having to mix synchronous and asynchronous calls and using limited APIs."

  This sounds like something that could potentially help with #3602 . It doesn't seem like there's anything we need to do about it that'd block the update, though.

**Comment by gnprice** [comment]
Merged #3658 and about to send it to beta for wider testing. Let's leave this open until we're done validating it, though.

In particular, over on that thread @ray-kraesig noted:
> on the Google Play Console there's a list of currently-greylisted APIs we're using. None of them seem to be on the (very short) [Pie blacklist](https://android.googlesource.com/platform/frameworks/base/+/pie-release/config/hiddenapi-force-blacklist.txt), but they should probably still tracked down and verified harmless before a full release.

**Comment by ray-kraesig** [comment]
There are ten weakly-deprecated APIs reported to be in use by our app. Of those, only one has made it onto the Android 10 blacklist.

**TL;DR:** We're probably good for now.

### Used by OkHttp:
- `Lcom/android/org/conscrypt/OpenSSLSocketImpl;->getAlpnSelectedProtocol()[B`
- `Lcom/android/org/conscrypt/OpenSSLSocketImpl;->setAlpnProtocols([B)V`
- `Ldalvik/system/CloseGuard;->get()Ldalvik/system/CloseGuard;`
- `Ldalvik/system/CloseGuard;->open(Ljava/lang/String;)V`
- `Ldalvik/system/CloseGuard;->warnIfOpen()V`

**Issue:** https://github.com/square/okhttp/issues/3980

OkHttp is used by React Native to implement (at least) `fetch()`. As can be seen in the linked issue, Square's been on top of this.

Note that `getAlpnSelectedProtocol` has been [fully blacklisted in Android 10](https://developer.android.com/about/versions/10/non-sdk-q), so upgrading to the *next* SDK target is blocked on
 - [ ] this getting fixed in OkHttp
 - [ ] that fix getting into React Native
 - [ ] us upgrading React Native to a version with the fix

But this is true of literally *every* consumer of React Native, so this is probably a low risk.


### Used by React Native:
- `Landroid/widget/ScrollView;->mScroller:Landroid/widget/OverScroller;`

**Issue:** https://github.com/facebook/react-native/issues/25120

Unfortunately, this does *not* appear to be being handled by upstream; the linked issue was locked by a bot as "stale".

The stack trace indicates that this is used by the [ScrollView](https://facebook.github.io/react-native/docs/scrollview) component; digging indicates that this (reflective access of a private member variable) was originally introduced as a hack in [this commit](https://github.com/facebook/react-native/commit/36ca1a078acbaba770c196847254e7a6e80c6918).

We (like other React Native users) already suffer degradation of functionality on platforms where ScrollView has been replaced and this member variable does not exist (e.g., as in  https://github.com/facebook/react-native/issues/9697). A formal blacklisting of this API would annoy but not significantly impair us.

### Used by Android Support library:
- `Landroid/view/View;->computeFitSystemWindows(Landroid/graphics/Rect;Landroid/graphics/Rect;)Z`
- `Landroid/view/View;->mAccessibilityDelegate:Landroid/view/View$AccessibilityDelegate;`
- `Landroid/view/ViewGroup;->makeOptionalFitsSystemWindows()V`

One assumes that Google will fix their own support libraries before it becomes relevant elsewhere.


### Unknown:
- `Landroid/content/Context;->bindServiceAsUser(Landroid/content/Intent;Landroid/content/ServiceConnection;ILandroid/os/Handler;Landroid/os/UserHandle;)Z`

The string `bindServiceAsUser` does not occur anywhere in our APK. This is *probably* (?) from the underlying Android-provided `WebView` implementation, as relatedly conjectured [here](https://stackoverflow.com/a/51989550), but the stack trace isn't strongly indicative of that.

**Comment by gnprice** [comment]
> There are ten weakly-deprecated APIs reported to be in use by our app. Of those, only one has made it onto the Android 10 blacklist.

Cool -- thanks for all these details!

I've just gone and filed an issue (#3665) for the *next* upgrade, so we have a place to record that information.

**Comment by ray-kraesig** [comment]
I think we can call this issue resolved.

One concern that has not yet been acted on is the change to initial element focus. This is important for a11y... but we don't really have a coherent a11y strategy or framework at the moment; so presently it's just a technical detail to be kept in mind when we develop them.

Some informal notes on what was considered, copied from my scratchpad, follow. (These are ordered according to their appearance in the [official changelog](https://developer.android.com/about/versions/pie/android-9.0-changes-28).)

----

* `FOREGROUND_SERVICE` doesn't apply; we don't use [Service.startForeground](https://developer.android.com/reference/android/app/Service.html#startForeground(int,%20android.app.Notification))
* we don’t care about device serial numbers
* we don’t have our own DNS resolution mechanism
* `isCleartextTrafficPermitted` was fixed preemptively in 1f55c035
  * note: release builds can no longer connect directly to dev servers
* we have one WebView and one process
* we should already disconnect/reconnect when we switch default networks
  * we actually just rely on RN for this
  * due to long-polling, we might have exactly one off-network request
* Apache? what’s that?

#### UI changes
* initially-focused-element story is unfortunately not great right now
* color formatting changes don't affect us due to cross-compat
* no `file://` MIME type assumptions here (yet, nor now ever)
* scrolling behavior appears identical in tests
* “Notifications from suspended apps” sounds like it matters…
  * … but “suspended” isn’t actually a documented app lifecycle state
  * (new behavior is probably an improvement anyway)
